### PR TITLE
Kubeadm resource requests for etcd and kube-proxy

### LINKED ISSUE #1095: kubeadm should configure resource request for critical kube-system component
## Versions

**kubeadm version**:
```
&version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:14:39Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```

## What happened?
Connections to etcd are timing out on my heavy loaded machines. My suspicion is that this happens due to not reserving a minimum amount of CPU time for the service. 

I checked the `kubectl describe node` output for the resource requests:
```
  Namespace                  Name                                                CPU Requests  CPU Limits  Memory Requests  Memory Limits
  ---------                  ----                                                ------------  ----------  ---------------  -------------
  kube-system                coredns-78fcdf6894-bxv7k                            100m (5%)     0 (0%)      70Mi (0%)        170Mi (2%)
  kube-system                coredns-78fcdf6894-kr8jv                            100m (5%)     0 (0%)      70Mi (0%)        170Mi (2%)
  kube-system                etcd-myhost.com                                     0 (0%)        0 (0%)      0 (0%)           0 (0%)
  kube-system                kube-apiserver-myhost.com                           250m (12%)    0 (0%)      0 (0%)           0 (0%)
  kube-system                kube-controller-manager-myhost.com                  200m (10%)    0 (0%)      0 (0%)           0 (0%)
  kube-system                kube-proxy-zvcpw                                    0 (0%)        0 (0%)      0 (0%)           0 (0%)
  kube-system                kube-scheduler-myhost.com                           100m (5%)     0 (0%)      0 (0%)           0 (0%)
  kube-system                weave-net-ccjj4                                     20m (1%)      0 (0%)      0 (0%)           0 (0%)
```

## What you expected to happen?
The request here is to configure some amount of CPU requests for the etcd and kube-proxy.

### PR DESCRIPTION
**Pull request by peterableda**
**What this PR does / why we need it**:
Kubeadm doesn't configure resources for the etcd and kube-proxy components. Without resource requests, these pods fall in the BestEffort qosClass which means that both the kernel's OOM killer and the K8s eviction manager starts killing these processes first. We should protect these critical components more and this change is a step in that direction. 
After the change, both pods have Burstable qosClass.

I also hit an etcd timeout issue before where we suspected that the root cause is an etcd starvation. 

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubeadm/issues/1095

**Special notes for your reviewer**:
I tried to be conservative while setting the requests for the components, checked my internal cluster and came up with these values. If there are better usage metrics available which we can use to determine good defaults, let me know and I can update the values.

**Release note**:
```
kubeadm now configures resource requests for the etcd and kube-proxy so they fall in the Bustable qosClass
```

### CODE REVIEW

**Comment by neolit123** [comment]
@peterableda thank you for the PR.

i remembered the linked issue discussion.
it's always tricky to hardcode such resource values that affect **all users** in the manifests.
having said that, we already have similar for core/kube DNS.

/ok-to-test
/kind bug
/assign @timothysc

**Comment by timothysc** [comment]
We can bump priority without capping.  I guess my only concern is we pick a default and it doesn't scale. 

/cc @kubernetes/sig-scalability-pr-reviews @kubernetes/sig-network-pr-reviews

**Comment by timothysc** [comment]
The more I think about it, it makes much more sense to make it an uninterruptable priority.  

@bsalamat - What is gke doing here?

**Comment by timothysc** [comment]
@shyamjvs Do you have any data on the values from scale tests?

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go [hunk 1]
```diff
@@ -77,6 +77,10 @@ spec:
       - name: kube-proxy
         image: {{ .Image }}
         imagePullPolicy: IfNotPresent
+        resources:
+          requests:
+            cpu: 100m
+            memory: 50Mi
```
Comment (wojtek-t) [inline cmd/kubeadm/app/phases/addons/proxy/manifests.go]: 50Mi is fine for small clusters, but definitely way too small for larger ones (or to be more specific for clusters with big enough number of services and their backends)

File: cmd/kubeadm/app/phases/etcd/local.go [hunk 2]
```diff
@@ -74,6 +74,7 @@ func GetEtcdPodSpec(cfg *kubeadmapi.InitConfiguration) v1.Pod {
 			cfg, kubeadmconstants.Etcd, 2379, cfg.CertificatesDir,
 			kubeadmconstants.EtcdCACertName, kubeadmconstants.EtcdHealthcheckClientCertName, kubeadmconstants.EtcdHealthcheckClientKeyName,
 		),
+		Resources: staticpodutil.ComponentResources("100m"),
```
Comment (wojtek-t) [inline cmd/kubeadm/app/phases/etcd/local.go]: in k8s on GCE we have two etcd events:
- the main one with 200m
- the events one with 100m

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go (hunk 1 above)
Comment (timothysc) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: @wojtek-t do you have an upper bound?

Comment (wojtek-t) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: In 5k-node clusters with 150k pods, each pod being a backend for some service I've seen things like 300MB IIRC.
But we probably also don't want to allocate that much in every cluster.

**Comment by timothysc** [comment]
I'm going to close this PR, but we're going to loop back on this problem in 1.14.

**Pull request closed without merging**